pkg/gardenlet/controller/extensions/seed_control.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1alpha1"
+	gardenextensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	gardenextensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/logger"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// FederatedSeedController is responsible for maintaining multiple federated Seeds' controllers
+type FederatedSeedController struct {
```

Comment:
I think it would be better if we named this struct ShootStateController  and then it has a field which is called federatedExtensionSyncControllerManager which takes care of starting and keeping track of the federated extension sync controllers

Reply:
My idea was this controller to start multiple controllers that have to run for the federated seeds  not only the ShootState sync controller.